docs/docs/dynamo-store/architecture.md

```diff
@@ -0,0 +1,117 @@
+---
+sidebar_position: 1
+---
+
+# DynamoStore Architecture
+
+When it comes to building an event store on top of a serverless NoSQL database
+it is imperative to apply mechanical sympathy. We've seen implementations of
+DynamoDB event stores using a single-event-per-row strategy. This would seem the
+obvious choice so let's start by explaining why it's not the greatest idea.
+
+## DynamoDB cost model
+
+A great article to understand DynamoDB's pricing model was written by Zach
+Charles
+[here](https://zaccharles.medium.com/calculating-a-dynamodb-items-size-and-consumed-capacity-d1728942eb7c).
+
+The gist of it is that you pay per 4KB for reads, and 1KB for writes, and
+transactional writes will cost you double. As such DynamoStore is designed to
+utilise GetItem and PutItem APIs where possible and only does a table scan and
+transactional write when absolutely necessary.
+
+With these constraints enumerated it should be clear why an event-per-row
+schema is not the greatest idea
+
+1. Reading a stream necessitates a scan in which you pay 0.5RUs at minimum for
+   each event in the stream
+2. Writing will always be transactional so you'll pay double
+
+## The architecture
+
+At heart DynamoDB is a document store and an event sourced system is composed
+of multiple streams. It therefore stands to reason that we should model the
+storage such that the stream is the document. This works great until we run into
+the 400KB max item size. To solve for this we will move the current events to
+another document once the document size reaches a configurable threshold (in
+events or bytes). Streams are generally short-lived so calving should be a rare
+occurrence. For streams where it's unavoidable we provide functionality to store
+a snapshot alongside the main stream document ensuring that the state of the
+stream can be reconstructed in at most 1 round trip.
+
+To summarise:
+
+- A stream is represented as one or more documents in the store (a Tip document
+  and 0 or more calved batches of events)
```

Comment:
maybe batches -> pages in this ? Damn, realizing I called the batches - why did I do that! (Answer is that before events in tip, each write made a batch, and commit is defintiely a bad term)